eventmachine gem fails to install solaris 64 bit

eventmachine keeps failing when installing on Solaris in 64 bit mode

I dont know how to invoke correctly the options at installation to see if it is possible to tell it to not use g++ but SolarisStudio CC which is more appopriate. 
So far I have been able to run some test locally in Solaris 11.2 with Solaris Studio and i can install if I add the following to the extconf.rb files, just dont know how to do on the fly

CONFIG["CC"] = "cc -m64"
CONFIG["CXX"] = “CC -m64"
CONFIG["CFLAGS"] = "-m64 -xO3 -xbuiltin=%all -xinline=auto -xprefetch=auto -xdepend -KPIC"
CONFIG["LDFLAGS"] = "-m64 -L. -L/usr/sfw/lib -R/usr/sfw/lib"
CONFIG["CPPFLAGS"] = "-I/usr/sfw/include -I/usr/include"
CONFIG["CPP"] = "cc -E"
CONFIG["CCDLFLAGS"] = “-KPIC"


dolly@vdocklab01:~$ CXX=CC ; export CXX ; gem install eventmachine
Building native extensions.  This could take a while...
ERROR:  Error installing eventmachine:
        ERROR: Failed to build gem native extension.

    /export/home/dolly/ruby64/ruby-2.1.6/bin/ruby extconf.rb
checking for rb_trap_immediate in ruby.h,rubysig.h... no
checking for rb_thread_blocking_region()... yes
checking for ruby/thread.h... yes
checking for rb_thread_call_without_gvl() in ruby/thread.h... yes
checking for inotify_init() in sys/inotify.h... no
checking for __NR_inotify_init in sys/syscall.h... no
checking for writev() in sys/uio.h... yes
checking for rb_thread_fd_select()... yes
checking for rb_fdset_t in ruby/intern.h... yes
checking for rb_wait_for_single_fd()... yes
checking for rb_enable_interrupt()... no
checking for rb_time_new()... yes
checking for sys/event.h... no
checking for main() in -lnsl... yes
checking for main() in -lsocket... yes
checking for clock_gettime()... yes
checking for CLOCK_MONOTONIC_RAW in time.h... no
checking for CLOCK_MONOTONIC in time.h... yes
creating Makefile

make "DESTDIR=" clean

make "DESTDIR="
compiling binder.cpp
g++: error: unrecognized command line option ‘-KPIC’
make: *** [binder.o] Error 1

make failed, exit code 2

--- a/ext/ed.cpp
+++ b/ext/ed.cpp
@@ -1024,6 +1024,8 @@ void ConnectionDescriptor::_WriteOutboundData()
 	for(int i = 0; i < iovcnt; i++){
 		OutboundPage *op = &(OutboundPages[i]);
 		#ifdef CC_SUNWspro
+		// TODO: The void * cast works fine on Solaris 11, but
+		// I don't know at what point that changed from older Solaris.
 		iov[i].iov_base = (char *)(op->Buffer + op->Offset);
 		#else
 		iov[i].iov_base = (void *)(op->Buffer + op->Offset);

--- a/ext/em.h
+++ b/ext/em.h
@@ -93,6 +93,19 @@ typedef fd_set rb_fdset_t;
   rb_thread_select(fd_check((n)-1) ? (n) : FD_SETSIZE, (rfds), (wfds), (efds), (timeout))
 #endif
 
+
+// This Solaris fix is adapted from eval_intern.h in Ruby 1.9.3:
+// Solaris sys/select.h switches select to select_large_fdset to support larger
+// file descriptors if FD_SETSIZE is larger than 1024 on 32bit environment.
+// But Ruby doesn't change FD_SETSIZE because fd_set is allocated dynamically.
+// So following definition is required to use select_large_fdset.
+#ifdef HAVE_SELECT_LARGE_FDSET
+#define select(n, r, w, e, t) select_large_fdset((n), (r), (w), (e), (t))
+extern "C" {
+  int select_large_fdset(int, fd_set *, fd_set *, fd_set *, struct timeval *);
+}
+#endif
+
 class EventableDescriptor;
 class InotifyDescriptor;
 struct SelectData_t;

--- a/ext/extconf.rb
+++ b/ext/extconf.rb
@@ -125,6 +125,10 @@ when /solaris/
   add_define 'OS_SOLARIS8'
   check_libs(%w[nsl socket], true)
 
+  # If Ruby was compiled for 32-bits, then select() can only handle 1024 fds
+  # There is an alternate function, select_large_fdset, that supports more.
+  add_define 'HAVE_SELECT_LARGE_FDSET' if have_func('select_large_fdset', 'sys/select.h')
+
   if CONFIG['CC'] == 'cc' && (
      `cc -flags 2>&1` =~ /Sun/ || # detect SUNWspro compiler
      `cc -V 2>&1` =~ /Sun/        # detect Solaris Studio compiler

--- a/ext/extconf.rb
+++ b/ext/extconf.rb
@@ -125,13 +125,17 @@ when /solaris/
   add_define 'OS_SOLARIS8'
   check_libs(%w[nsl socket], true)
 
-  if CONFIG['CC'] == 'cc' and `cc -flags 2>&1` =~ /Sun/ # detect SUNWspro compiler
+  if CONFIG['CC'] == 'cc' && (
+     `cc -flags 2>&1` =~ /Sun/ || # detect SUNWspro compiler
+     `cc -V 2>&1` =~ /Sun/        # detect Solaris Studio compiler
+    )
     # SUN CHAIN
     add_define 'CC_SUNWspro'
     $preload = ["\nCXX = CC"] # hack a CXX= line into the makefile
     $CFLAGS = CONFIG['CFLAGS'] = "-KPIC"
     CONFIG['CCDLFLAGS'] = "-KPIC"
     CONFIG['LDSHARED'] = "$(CXX) -G -KPIC -lCstd"
+    CONFIG['LDSHAREDXX'] = "$(CXX) -G -KPIC -lCstd"
   else
     # GNU CHAIN
     # on Unix we need a g++ link, not gcc.

--- a/ext/fastfilereader/extconf.rb
+++ b/ext/fastfilereader/extconf.rb
@@ -56,13 +56,17 @@ when /solaris/
   add_define 'OS_SOLARIS8'
   check_libs(%w[nsl socket], true)
 
-  if CONFIG['CC'] == 'cc' and `cc -flags 2>&1` =~ /Sun/ # detect SUNWspro compiler
+  if CONFIG['CC'] == 'cc' && (
+     `cc -flags 2>&1` =~ /Sun/ || # detect SUNWspro compiler
+     `cc -V 2>&1` =~ /Sun/        # detect Solaris Studio compiler
+    )
     # SUN CHAIN
     add_define 'CC_SUNWspro'
     $preload = ["\nCXX = CC"] # hack a CXX= line into the makefile
     $CFLAGS = CONFIG['CFLAGS'] = "-KPIC"
     CONFIG['CCDLFLAGS'] = "-KPIC"
     CONFIG['LDSHARED'] = "$(CXX) -G -KPIC -lCstd"
+    CONFIG['LDSHAREDXX'] = "$(CXX) -G -KPIC -lCstd"
   else
     # GNU CHAIN
     # on Unix we need a g++ link, not gcc.

--- a/ext/fastfilereader/mapper.cpp
+++ b/ext/fastfilereader/mapper.cpp
@@ -89,6 +89,8 @@ void Mapper_t::Close()
 	// Calls to GetChunk are invalid after a call to Close.
 	if (MapPoint) {
 		#ifdef CC_SUNWspro
+		// TODO: The void * cast works fine on Solaris 11, but
+		// I don't know at what point that changed from older Solaris.
 		munmap ((char*)MapPoint, FileSize);
 		#else
 		munmap ((void*)MapPoint, FileSize);

--- a/tests/em_test_helper.rb
+++ b/tests/em_test_helper.rb
@@ -41,6 +41,10 @@ class Test::Unit::TestCase
       RbConfig::CONFIG['host_os'] =~ /mswin|mingw/
     end
 
+    def solaris?
+      RUBY_PLATFORM =~ /solaris/
+    end
+
     # http://stackoverflow.com/questions/1342535/how-can-i-tell-if-im-running-from-jruby-vs-ruby/1685970#1685970
     def jruby?
       defined? JRUBY_VERSION

--- a/tests/test_file_watch.rb
+++ b/tests/test_file_watch.rb
@@ -34,6 +34,7 @@ class TestFileWatch < Test::Unit::TestCase
     end
 
     def test_events
+      omit_if(solaris?)
       EM.run{
         file = Tempfile.new('em-watch')
         $tmp_path = file.path

